packages/stylist-core/src/ast/rule.rs

```diff
@@ -22,24 +20,25 @@ pub struct Rule {
     pub condition: Cow<'static, [StringFragment]>,
     /// Note that not all At-Rules allow arbitrary other At-Rules to appear
     /// inside them, or arbitrary blocks. No safeguards at this point!
-    pub content: Cow<'static, [RuleContent]>,
+    pub content: Cow<'static, [RuleBlockContent]>,
 }
 
 impl ToStyleStr for Rule {
-    fn write_style<W: fmt::Write>(&self, w: &mut W, class_name: Option<&str>) -> Result<()> {
+    fn write_style(&self, w: &mut String, ctx: &mut StyleContext<'_>) {
+        let mut cond = "".to_string();
         for frag in self.condition.iter() {
-            frag.write_style(w, class_name)?;
+            frag.write_style(&mut cond, ctx);
         }
 
-        writeln!(w, " {{")?;
+        let mut rule_ctx = ctx.with_rule_condition(&cond);
+        if cond.starts_with("@keyframes") {
+            rule_ctx.start(w); // keyframes should always be printed.
+        }
```

Comment:
This is to prevent printing an excessive empty blocks when entering and exiting a rule inside a block.

Consider the following stylesheet:

```css
header, footer {
    border: 1px solid black;

    @supports (max-width: 500px) {
        @media screen and (max-width: 500px) {
            display: flex;
        }
    }
}
```

When they are always started, it will print the following stylesheet:

```css
.stylist-abcdefgh header, .stylist-abcdefgh footer {
    border: 1px solid black;
}
@supports (max-width: 500px) {
    .stylist-abcdefgh header, .stylist-abcdefgh footer {
        /* empty block */
    }
}
@supports (max-width: 500px) {
    @media screen and (max-width: 500px) {
        .stylist-abcdefgh header, .stylist-abcdefgh footer {
            display: flex;
        }
    }
}
```

When lazily started (started in `StyleAttr`), it prints the following:

```css
.stylist-abcdefgh header, .stylist-abcdefgh footer {
    border: 1px solid black;
}
@supports (max-width: 500px) {
    @media screen and (max-width: 500px) {
        .stylist-abcdefgh header, .stylist-abcdefgh footer {
            display: flex;
        }
    }
}
```

However, for `@keyframes`, it always need to be printed as it has a name.

Reply:
:+1: You could write this as `let introduces_global_name = cond.starts_with("@keyframes"); if introduces_global_name` so it's clear that potentially more rules have to be considered going forward.